### ISSUE TITLE
Mapped type broke in 3.3: Type is not assignable to 'string | number | symbol'

### ISSUE THREAD

**Issue by m93a** [Microsoft/TypeScript#30331]
**TypeScript Version:**
 * broken in: 3.3.3333
 * works in: 3.1.4

**Search Terms:**
Mapped type key is wrong type. Can't use computed/generic type as key in mapped type. Infered type not working as key even if in union with `keyof`.

**Code**
```ts
type a<T> = T extends Array<infer U> ? U : never;
type b<T> = { [K in a<T> & keyof T ]: 42 };
```

**Expected behavior:**
Throws no error.

**Actual behavior:**
```
Type 'a<T> & keyof T' is not assignable to type 'string | number | symbol'.
    Type 'a<T> & keyof T' is not assignable to type 'symbol'.
```

**Workaround:**
```ts
type b<T> = { [K in a<T> & keyof T & keyof object]: 42 };
```

**Playground Link:** [here](https://www.typescriptlang.org/play/#src=type%20a%3CT%3E%20%3D%20T%20extends%20Array%3Cinfer%20U%3E%20%3F%20U%20%3A%20never%3B%0D%0Atype%20b%3CT%3E%20%3D%20%7B%20%5BK%20in%20a%3CT%3E%20%26%20keyof%20T%20%5D%3A%2042%20%7D%3B)

**Comment by jack-williams** [comment]
So looking into this I think the regression was probably caused by the constraint of the conditional type `a<T>` going from a union to the empty object type `{}`. The issue appears to be this part of `isRelatedTo`.
```ts
if (!result && source.flags & TypeFlags.Intersection) {
    // The combined constraint of an intersection type is the intersection of the constraints of
    // the constituents. When an intersection type contains instantiable types with union type
    // constraints, there are situations where we need to examine the combined constraint. One is
    // when the target is a union type. Another is when the intersection contains types belonging
    // to one of the disjoint domains. For example, given type variables T and U, each with the
    // constraint 'string | number', the combined constraint of 'T & U' is 'string | number' and
    // we need to check this constraint against a union on the target side. Also, given a type
    // variable V constrained to 'string | number', 'V & number' has a combined constraint of
    // 'string & number | number & number' which reduces to just 'number'.
    const constraint = getUnionConstraintOfIntersection(<IntersectionType>source, !!(target.flags & TypeFlags.Union));
    if (constraint) {
        if (result = isRelatedTo(constraint, target, reportErrors, /*headMessage*/ undefined, isIntersectionConstituent)) {
            errorInfo = saveErrorInfo;
        }
    }
}
```

The union constraint of `a<T> & keyof T` is coming back undefined because the constraint of `a<T>` is not a union. See the code in `getUnionConstraintOfIntersection`
```ts
if (constraint) {
    // A constraint that isn't a union type implies that the final type would be a non-union
    // type as well. Since non-union constraints are of no interest, we can exit here.
    if (!(constraint.flags & TypeFlags.Union)) {
        return undefined;
    }
```
Is this a valid thing to do? If the constraint of `a<T>` is `{}` and the constraint of `keyof T` is `string | number | symbol`, I would expect the combined constraint to be the intersection of the two.

**Comment by ahejlsberg** [comment]
> Is this a valid thing to do? If the constraint of a<T> is {} and the constraint of keyof T is string | number | symbol, I would expect the combined constraint to be the intersection of the two.

You're right, that doesn't seem like the right thing to do. I will remove that check.

**Comment by philraj** [comment]
This is still broken in 3.5.2 (and 3.5.1, [click here](https://www.typescriptlang.org/play/#code/C4TwDgpgBAKhDOwoF4oG8BQVtQIYC4oA7AVwFsAjCAJwG4scKB+Q0ym+gXww1EigCyuMJAAmAHhgA+FOgbYA2gAUoASyKwAuoRjLNXIA)). 

Was the fix released in 3.5.0?
